pkg/progressreader/progressstatus.go

```diff
@@ -0,0 +1,72 @@
+package progressreader
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"github.com/docker/docker/vendor/src/github.com/Sirupsen/logrus"
+)
+
+type ProgressStatus struct {
+	sync.Mutex
+	c         chan struct{}
+	observers []io.Writer
+	history   bytes.Buffer
+}
+
+func NewProgressStatus() *ProgressStatus {
+	return &ProgressStatus{
+		c:         make(chan struct{}),
+		observers: []io.Writer{},
+	}
+}
+
+func (ps *ProgressStatus) Write(p []byte) (n int, err error) {
+	ps.Lock()
+	defer ps.Unlock()
+	ps.history.Write(p)
+	for _, w := range ps.observers {
+		// copy paste from MultiWriter, replaced return with continue
+		n, err = w.Write(p)
+		if err != nil {
+			continue
+		}
+		if n != len(p) {
+			err = io.ErrShortWrite
+			continue
+		}
+	}
+	return len(p), nil
+}
+
+func (ps *ProgressStatus) AddObserver(w io.Writer) {
+	ps.Lock()
+	defer ps.Unlock()
+	w.Write(ps.history.Bytes())
+	ps.observers = append(ps.observers, w)
+}
+
+func (ps *ProgressStatus) Done() {
+	ps.Lock()
```

Comment:
I agree with you in principle. I just looked into using a condition variable instead of a channel, but found that a channel was actually better suited to this. A read from a closed channel will finish immediately, but a `Wait()` on a condition variable will always block. That means with a condition variable, we'd need some extra logic to prevent a race between `Wait` and `Done`. So after playing with the alternative, I think this approach is actually very reasonable.